experimental/InjectiveResolutions/src/MonoidAlgebra.jl

```diff
@@ -0,0 +1,611 @@
+########################################################################
+# MonoidAlgebra.jl
+#
+# This is a wrapper type for `MPolyRing` or `MPolyQuoRing` which is 
+# graded and available at the field `.algebra`. This datum is enhanced 
+# by various combinatorial stuff. 
+#
+# In order to facilitate functionality for ideals, we introduce a 
+# wrapper type `MonoidAlgebraIdeal` for `MPolyIdeal` (resp. 
+# `MPolyQuoIdeal`). For finitely generated modules, on the other hand, 
+# we implement the translation to the singular side within the 
+# `ModuleGens` directly; see `ModuleFunctionality.jl` for that. 
+########################################################################
+
+struct FaceQ # face of semigroup  
+  prime::Union{MPolyIdeal,MPolyQuoIdeal} #homogeneous prime corresponding to face
+  poly::Polyhedron  #face as polyhedron
+end
+
+struct HyperplaneQ # a hyperplane bounding the cone RR_{\geq 0}Q
+  hyperplane::Polyhedron
+  A::Matrix{Int}
+  b::Vector{Int}
+end
+@attributes mutable struct MonoidAlgebra{CoeffType,AlgebraType} <: Ring # monoid algebra with associated data
+  algebra::AlgebraType
+  pointed::Union{Nothing,Bool} #cone pointed
+  polyhedral_cone::Cone{QQFieldElem}
+  cone::Polyhedron
+  faces::Vector{FaceQ}
+  hyperplanes::Vector{HyperplaneQ}
+  zonotope::Tuple{Polyhedron,Vector{Int}}
+
+  function MonoidAlgebra(
+    A::AlgebraType; check::Bool=true
+  ) where {AlgebraType<:Union{MPolyRing,MPolyQuoRing}}
+    #check if monoid algebra
+    @check is_zm_graded(A) "given algebra is not ZZ^d-graded"
+    gg_Q = grading_group(A)
+    @check is_free(gg_Q) && is_abelian(gg_Q) "given algebra is not a monoid algebra"
+    kk = coefficient_ring(A)
+    result = new{elem_type(kk),AlgebraType}(A, nothing)
+  end
+end
+
+is_graded(A::MonoidAlgebra) = true
+is_zm_graded(A::MonoidAlgebra) = true
+
+function polyhedral_cone(A::MonoidAlgebra)
+  if !isdefined(A, :polyhedral_cone)
+    A.polyhedral_cone = get_polyhedral_cone(A.algebra)
+  end
+  return A.polyhedral_cone
+end
```

Comment:
If you don't need every nanosecond, you could also handle this with an attribute and subsequently also get rid of the `MonoidAlgebra::polyhedral_cone` field:
```suggestion
@attr Cone{QQFieldElem} polyhedral_cone(A::MonoidAlgebra) = get_polyhedral_cone(A.algebra)
```
The same of course holds for `cone`, `faces`, `hyperplanes`, ... (for each this could be asked separately of course)